cmd/google_cloud_ops_agent_diagnostics/main_windows.go

```diff
@@ -0,0 +1,120 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build windows
+// +build windows
+
+package main
+
+import (
+	"flag"
+	"fmt"
+	"os"
+
+	"github.com/GoogleCloudPlatform/ops-agent/internal/self_metrics"
+	"golang.org/x/sys/windows/svc"
+	"golang.org/x/sys/windows/svc/debug"
+	"golang.org/x/sys/windows/svc/eventlog"
+)
+
+const (
+	eventID                 uint32 = 1
+	ERROR_SUCCESS           uint32 = 0
+	ERROR_FILE_NOT_FOUND    uint32 = 2
+	ERROR_INVALID_DATA      uint32 = 13
+	ERROR_INVALID_PARAMETER uint32 = 87
+)
+
+type service struct {
+	log      debug.Log
+	userConf string
+}
+
+func run() error {
+	name := "google-cloud-ops-agent-diagnostics"
+	elog, err := eventlog.Open(name)
```

Comment:
I don't think it was mentioned in the design doc, so is there any specific reason for why we decided to log to the eventlog rather than logging to a file? Does this have the same limitations as with otel, where certain panics are ommited?

Reply:
Good call! I will add details to the design doc. The implementation of the `Google Cloud Ops Agent - Diagnostics` service in Window is mirroring the current implementation of the main service that manages the subagents in windows. This way our logging will appear in the same places, journald in Linux and eventlog in Windows, as the main service. 

https://github.com/GoogleCloudPlatform/ops-agent/blob/6f7c78196f993c6eadc12101ec9154e36b5c9773/cmd/ops_agent_windows/run_windows.go#L46-L54

I wasn't completely aware of the limitation of omitting panics in the `opentelemetry-collector` subagent, but considering my current familiarity of using the `opentelemetry sdk` (which shares some code), I can say this would work similarly. We are catching and logging errors in every step of setting the metric collection pipeline, but (probably) `panics` are handled internally and not going to be logged. Have to ask @dashpole more about this.

Reply:
Are you asking if the OTel SDK will ever panic?  That should never happen per [the OTel specification](https://github.com/open-telemetry/opentelemetry-specification/blob/27b50425454de9efc56b43e54767a581bdada71c/specification/error-handling.md#basic-error-handling-principles): `The SDK MUST NOT throw unhandled exceptions for errors in their own operations. For example, an exporter should not throw an exception when it cannot reach the endpoint to which it sends telemetry data.`. If it does it is a bug in the SDK :)

Reply:
> I don't think it was mentioned in the design doc, so is there any specific reason for why we decided to log to the eventlog rather than logging to a file? Does this have the same limitations as with otel, where certain panics are ommited?

@hsmatulisgoogle are you thinking about some customer cases we had that the error logs were swallowed without us knowing the real causes? Logging into the eventlog is not the reason why those error logs were swallowed. Logging into the eventlog actually prevents the swallowing.  Using [log](https://github.com/GoogleCloudPlatform/opentelemetry-operations-collector/blob/master/cmd/otelopscol/main.go#L33) and windows service together is the actual reason that leads to uncaptured logs.

Reply:
That's right, thanks for clarifying sophie. That is what i was worried about but it seems like it is not an issue

Reply:
Would a panic be captured in the windows event log as things are?

Reply:
Making a summary of all the information in this thread :

1. The OTel SDK won't ever `panic` (or throw any unhandled exception) by design . If it does, it is a bug upstream that should be fixed and it won't be logged anywhere.
2. All the errors thrown by the collection and / or sending of metrics are caught and reported accordingly in the `self_metrics.CollectOpsAgentSelfMetrics` method. (well, that is the expectation of the final code of this PR)
3. The resulting errors are going to be sent to the event log of the `Diagnostics Service`
https://github.com/GoogleCloudPlatform/ops-agent/blob/f60954478217da5ce4756c796f8c3f077b4c525c/cmd/google_cloud_ops_agent_diagnostics/main_windows.go#L101-L107

Reply:
@hsmatulisgoogle `Would a panic be captured in the windows event log as things are?` 

If there was a `panic`, it wouldn't be captured into the windows event log, but the guarantee of the OTel SDK is that it won't `panic` at all.

Reply:
Spoke about it offline, but yeah that's what I was thinking!

Reply:
Added deferred recovery in https://github.com/GoogleCloudPlatform/ops-agent/pull/852/commits/2ba114c3be67cdda4dece6ab237106dd032f3757 .